File: src/main/java/io/protobase/f7/formulas/math/TANH.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class TANH extends AbstractFormula {
  public static FormulaName NAME = FormulaName.TANH;
  public static TANH SELF = new TANH();

  public TANH() {
    super();
  }

  public TANH(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 1, NAME);
    return Math.tanh(Converters.toDouble(Converters.first(collateralLookup.apply(origin, values[0]))));
  }
}

File: src/main/js/formulas/math/TANH.ts
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";

export class TANH extends AbstractFormula {
  static SELF: TANH = new TANH();
  NAME = FormulaName.TANH;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 1, this.NAME);
    return Math["tanh"](
      Converters.toNumber(Converters.first(this.collateralLookup(origin, values[0])))
    );
  }
}
